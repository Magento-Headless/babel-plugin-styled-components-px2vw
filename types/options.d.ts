export interface IPx2VwOptions {
  readonly unitToConvert: string;
  readonly viewportWidth: number;
  readonly unitPrecision: number;
  readonly viewportUnit: string;
  readonly fontViewportUnit: string;
  readonly selectorBlackList: ReadonlyArray<string>;
  readonly propList: ReadonlyArray<string>;
  readonly minPixelValue: number;
  readonly mediaQuery: boolean;
  readonly exclude?: RegExp | ReadonlyArray<RegExp>;
  readonly include?: RegExp | ReadonlyArray<RegExp>;
  readonly replace: boolean;
  readonly landscape: boolean;
  readonly landscapeUnit: string;
  readonly landscapeWidth: number;
}
