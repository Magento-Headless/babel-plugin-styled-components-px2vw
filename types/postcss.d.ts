declare module 'postcss-px-to-viewport' {
  import { Plugin } from 'postcss';
  import { IPx2VwOptions } from 'types/options';
  // import { PluginCreator } from 'postcss'; # postcss v8

  const px2vw: Plugin<IPx2VwOptions>;

  export = px2vw;
}
