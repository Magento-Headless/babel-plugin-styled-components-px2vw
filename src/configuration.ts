import { IPx2VwOptions } from '../types/options';
// type TPx2Vw = typeof Px2VwPlugin;
// export type Px2VwPluginOptions = Parameters<TPx2Vw>;
export type IConfiguration = IPx2VwOptions & {
  tags: ReadonlyArray<string>;
};
class ConfigurationManager {
  private static readonly defaultConfiguration: IConfiguration = {
    unitToConvert: 'px',
    viewportWidth: 750,
    unitPrecision: 5,
    propList: ['*'],
    viewportUnit: 'vw',
    fontViewportUnit: 'vw',
    selectorBlackList: [],
    minPixelValue: 1,
    mediaQuery: false,
    replace: true,
    exclude: undefined,
    include: undefined,
    landscape: false,
    landscapeUnit: 'vw',
    landscapeWidth: 568,
    tags: ['styled', 'css', 'createGlobalStyle', 'keyframes'],
  };
  private _config: IConfiguration = ConfigurationManager.defaultConfiguration;

  public get config(): IConfiguration {
    return this._config;
  }

  public updateConfig(config?: Partial<IConfiguration>): void {
    if (config) {
      this._config = Object.assign({}, ConfigurationManager.defaultConfiguration, config);
    }
  }
}

export default new ConfigurationManager();
