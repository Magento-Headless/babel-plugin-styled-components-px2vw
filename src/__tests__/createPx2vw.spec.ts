import { identifier } from '@babel/types';
import generate from '@babel/generator';
import { runInNewContext } from 'vm';
import createPx2vw from '../createPx2vw';
import configuration, { IConfiguration } from '../configuration';

function px2vw(
  value: string,
  {
    viewportWidth = configuration.config.viewportWidth,
    unitPrecision = configuration.config.unitPrecision,
    minPixelValue = configuration.config.minPixelValue,
  }: Partial<IConfiguration> = configuration.config,
): string {
  const sandbox = { result: '' };
  const ast = createPx2vw(identifier('px2vw'), { viewportWidth, unitPrecision, minPixelValue });
  const code = generate(ast).code;
  runInNewContext(`${code} result = px2vw(${value});`, sandbox);
  return sandbox.result;
}

it('should match snapshot', function () {
  const ast = createPx2vw(identifier('px2vw'), configuration.config);
  expect(generate(ast).code).toMatchSnapshot();
});

it('should transform String', function () {
  expect(px2vw("'-100px'")).toBe('-13.33333vw');
  expect(px2vw("'32px'")).toBe('4.26667vw');
  expect(px2vw("'11.3333333px'")).toBe('1.51111vw');
  expect(px2vw("'-11.3333333px'")).toBe('-1.51111vw');
});

it('should transform Number', function () {
  expect(px2vw('16')).toBe('2.13333vw');
  expect(px2vw('16.3333333')).toBe('2.17778vw');
  expect(px2vw('-16')).toBe('-2.13333vw');
  expect(px2vw('-16.3333333')).toBe('-2.17778vw');
});

it('should transform Function', function () {
  expect(px2vw('function() {return 20;}')).toBe('2.66667vw');
  expect(px2vw('function(props) { return props.width / 2; }, { width: 100 }')).toBe('6.66667vw');
});

// it('should ignore value less than "minPixelValue" option', function () {
//   const options = { minPixelValue: 2 };
//   expect(px2vw("'1px'", options)).toBe('1px');
//   expect(px2vw('1', options)).toBe('1px');
//   expect(px2vw("'1.3333333px'", options)).toBe('1.3333333px');
//   expect(px2vw('1.3333333', options)).toBe('1.3333333px');
//   expect(px2vw("'-1px'", options)).toBe('-1px');
//   expect(px2vw('-1', options)).toBe('-1px');
//   expect(px2vw("'-1.3333333px'", options)).toBe('-1.3333333px');
//   expect(px2vw('-1.3333333', options)).toBe('-1.3333333px');
//   expect(px2vw('function() {return 1;}', options)).toBe('1px');
//   expect(px2vw('function() {return "1px";}', options)).toBe('1px');
//   expect(px2vw('function(props) { return props.width / 2; }, { width: 2 }', options)).toBe('1px');
//   expect(px2vw('function(props) { return props.width / 2 + "px"; }, { width: 2 }', options)).toBe('1px');
// });

// it('should multiply "multiplier" option', function () {
//   const options = { minPixelValue: 2, multiplier: 2 };
//   expect(px2vw("'1px'", options)).toBe('1px');
//   expect(px2vw('1', options)).toBe('1px');
//   expect(px2vw("'1.3333333px'", options)).toBe('1.3333333px');
//   expect(px2vw('1.3333333', options)).toBe('1.3333333px');
//   expect(px2vw("'-1px'", options)).toBe('-1px');
//   expect(px2vw('-1', options)).toBe('-1px');
//   expect(px2vw("'-1.3333333px'", options)).toBe('-1.3333333px');
//   expect(px2vw('-1.3333333', options)).toBe('-1.3333333px');
//   expect(px2vw('function() {return 1;}', options)).toBe('1px');
//   expect(px2vw('function() {return "1px";}', options)).toBe('1px');
//   expect(px2vw('function(props) { return props.width / 2; }, { width: 2 }', options)).toBe('1px');
//   expect(px2vw('function(props) { return props.width / 2 + "px"; }, { width: 2 }', options)).toBe('1px');
//   expect(px2vw("'2px'", options)).toBe('0.04rem');
//   expect(px2vw('2', options)).toBe('0.04rem');
//   expect(px2vw("'2.3333333px'", options)).toBe('0.04667rem');
//   expect(px2vw("'-2.3333333px'", options)).toBe('-0.04667rem');
//   expect(px2vw('2.3333333', options)).toBe('0.04667rem');
//   expect(px2vw('-2.3333333', options)).toBe('-0.04667rem');
//   expect(px2vw("'-2px'", options)).toBe('-0.04rem');
//   expect(px2vw('-2', options)).toBe('-0.04rem');
//   expect(px2vw('function() {return 2;}', options)).toBe('0.04rem');
//   expect(px2vw('function() {return "2px";}', options)).toBe('0.04rem');
//   expect(px2vw('function(props) { return props.width / 2; }, { width: 4 }', options)).toBe('0.04rem');
//   expect(px2vw('function(props) { return props.width / 2 + "px"; }, { width: 4 }', options)).toBe('0.04rem');
// });
