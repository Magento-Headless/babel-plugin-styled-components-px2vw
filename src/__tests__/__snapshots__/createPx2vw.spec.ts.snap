// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
"function px2vw(input, ...args) {
  if (typeof input === 'function') return px2vw(input(...args), ...args);
  var value = typeof input === 'string' ? parseFloat(input) : typeof input === 'number' ? input : 0;
  var pixels = Number.isNaN(value) ? 0 : value;

  if (Math.abs(pixels) < 1) {
    return \`\${pixels}px\`;
  }

  var mul = Math.pow(10, 5 + 1);
  return \`\${Math.round(Math.floor(pixels * 100 / 750 * mul) / 10) * 10 / mul}vw\`;
}"
`;
